enhancements/cvo/object-removal-manifest-annotation.md

```diff
@@ -0,0 +1,142 @@
+---
+title: use-manifest-annotation-for-object-removal
+authors:
+  - "@jottofar"
+reviewers:
+  - "@LalatenduMohanty"
+  - "@wking"
+  - "@crawford"
+  - "@sdodson"
+approvers:
+  - "@crawford"
+  - "@sdodson"
+creation-date: 2020-08-03
+last-updated: 2020-08-04
+status: implementable
+---
+
+# Use Manifest Annotation For Object Removal
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Currently CVO managed object removal is handled by jobs.
+Jobs have many immutable components, so they aren't a great match for CVO managed manifests.
+Instead of the current approach manifests will be used by adding a new annotation to a manifest which will be handled by the CVO causing it to delete the in-cluster object rather than creating/reconciling the manifest.
+This will provide a more straightforward way to remove content.
+
+## Motivation
+
+There should be a straightforward way to remove CVO managed objects.
+
+### Goals
+
+* Users will be able to remove any of the currently managed CVO objects by simply modifying the object's exisitng manifest and adding the new delete annotation.
+
+### Non-Goals
+
+* This enhancement does not add any new objects for managment by CVO therefore the new delete annotation only applies to the currently supported manifests and any that may be added in the future.
+
+## Proposal
+
+When the following annotation appears in a CVO supported manifest and is set to "true" the associated object will be removed from the cluster by the CVO:
```

Comment:
Do we want to define the behavior of other values?  If nothing else, I'd like to explicitly call them out as [undefined][1], with wording like:

> Values other than `true` have undefined behavior.

Possibly linking to Wikipedia or other spec source that distinguishes between "unspecified" and "undefined".

[1]: https://en.wikipedia.org/wiki/Undefined_behavior

Reply:
I think simply saying "Values other than true will be ignored." is clear enough. I do say that "false" will result in a warning log which should be changed to "any value other than true".

Reply:
Changed in commit [262765d](https://github.com/openshift/enhancements/pull/419/commits/262765d0e846f11bb32d17fe4841e5b9917d4a82)

Reply:
> Values other than `true` will be ignored.

This is pretty soft, and I'd rather it be unambiguously a disaster if someone typos in an undefined value.  We don't want to accidentally keep applying a manifest that the operator maintainers intended to be a for-deletion marker.

Reply:
So if we find an undefined value we want to ignore the entire manifest, correct? And probably log an error.
However if we find `false` that would just be a no-op and we'd process the manifest as normal. Perhaps put out a warning just in case the intent was to delete and they mistakenly set it to false?

Reply:
I think `LoadUpdate` should error out if it finds a non-`true` value, so the CVO would fail before getting as far as manifest reconciliation.  That way there's no risk that the broken PR passes its presubmits and lands.

Reply:
Changed in commit [39e010c](https://github.com/openshift/enhancements/pull/419/commits/39e010c5721d73d80cc39cd2dc5fb6268ccb7a25)